samcli/commands/init/__init__.py

```diff
@@ -39,107 +55,57 @@
-    build_msg = """
-Project generated: {output_dir}/{name}
-
-Steps you can take next within the project folder
-===================================================
-[*] Install dependencies
-[*] Invoke Function: sam local invoke HelloWorldFunction --event event.json
-[*] Start API Gateway locally: sam local start-api
-""".format(
-        output_dir=output_dir, name=name
-    )
-
-    no_build_step_required = (
-        "python",
-        "python3.7",
-        "python3.6",
-        "python2.7",
-        "nodejs",
-        "nodejs4.3",
-        "nodejs6.10",
-        "nodejs8.10",
-        "nodejs10.x",
-        "ruby2.5",
-    )
-    next_step_msg = no_build_msg if runtime in no_build_step_required else build_msg
-
-    try:
-        generate_project(location, runtime, dependency_manager, output_dir, name, no_input)
-        if not location:
-            click.secho(next_step_msg, bold=True)
-            click.secho("Read {name}/README.md for further instructions\n".format(name=name), bold=True)
-        click.secho("[*] Project initialization is now complete", fg="green")
-    except GenerateProjectFailedError as e:
-        raise UserException(str(e))
+def cli(ctx, no_interactive, location, runtime, dependency_manager, output_dir, name, app_template, no_input):
+    do_cli(
+        ctx, no_interactive, location, runtime, dependency_manager, output_dir, name, app_template, no_input
+    )  # pragma: no cover
+
+
+def do_cli(
+    ctx,
+    no_interactive,
+    location,
+    runtime,
+    dependency_manager,
+    output_dir,
+    name,
+    app_template,
+    no_input,
+    auto_clone=True,
+):
+    # check for mutually exclusive parameters
+    if location and app_template:
+        msg = """
+You must not provide both the --location and --app-template parameters.
+
+You can run 'sam init' without any options for an interactive initialization flow, or you can provide one of the following required parameter combinations:
+    --name and --runtime and --app-template
+    --location
+        """
+        raise UserException(msg)
+    # check for required parameters
+    if location or (name and runtime and dependency_manager and app_template):
+        # need to turn app_template into a location before we generate
+        extra_context = None
+        if app_template:
+            templates = InitTemplates(no_interactive, auto_clone)
+            location = templates.location_from_app_template(runtime, dependency_manager, app_template)
+            no_input = True
+            extra_context = {"project_name": name, "runtime": runtime}
+        if not output_dir:
+            output_dir = "."  # default - should I lift this to overall options and seed default?
+        do_generate(location, runtime, dependency_manager, output_dir, name, no_input, extra_context)
+    elif no_interactive:
+        error_msg = """
+ERROR: Missing required parameters, with --no-interactive set.
+
+Must provide one of the following required parameter combinations:
```

Comment:
this is not clear on what constitues a combination, I would put valid combinations in a list, so that its easy to grok for the user.

Reply:
There are only two, what format did you suggest?

Reply:
[--name--runtime , --dependency-manager, --app-template] or [--location]. Just having all the `ands` together in `--name and --runtime and --dependency-manager and --app-template` made it difficult for me to spot the combination.